### PR TITLE
feat: Mashmap

### PR DESCRIPTION
**Pull request by tdayris**
### Description

This PR adds [MashMap](https://github.com/marbl/MashMap), a long read aligner.

### QC

For all wrappers added by this PR, I made sure that

* [X] there is a test case which covers any introduced changes,
* [X] `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* [X] either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* [X] rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* [X] all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* [X] wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* [X] all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* [X] `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* [X] temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* [X] the `meta.yaml` contains a link to the documentation of the respective tool or command,
* [X] `Snakefile`s pass the linting (`snakemake --lint`),
* [X] `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* [X] Python wrapper scripts are formatted with [black](https://black.readthedocs.io).

### CODE REVIEW

File: bio/mashmap/environment.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+channels:
+  - bioconda
+  - conda-forge
+  - defaults
+dependencies:
+  - mashmap ==2.0
+  - gsl ==2.7
```
Comment (fgvieira) [inline bio/mashmap/environment.yaml:7]:
```suggestion
  - mashmap =2.0
  - gsl =2.7

```

File: bio/mashmap/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  - query: Path to query file (fasta, fastq)
+output:
+  - Path to the alignment file
+params:
+  - extra: Optional parameters for MashMap
+notes: |
+  `input.ref` may be either a path to a file containing sequences (fasta) or a text file containing a list
+  paths to fasta-formatted sequence files.
+  
+  `input.query` may be either a fastq file or a fasta file.
```
Comment (fgvieira) [inline bio/mashmap/meta.yaml:17]:
```suggestion
  * `input.ref` may be either a path to a fasta file or a text file containing a list of paths to several fasta files.
  * `input.query` may be either a fastq file or a fasta file.
```

File: bio/mashmap/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+__license__ = "MIT"
+
+from snakemake.shell import shell
+
+log = snakemake.log_fmt_shell(stdout=True, stderr=True)
+extra = snakemake.params.get("extra", "")
+
+# Handling input file types (either a fasta file, or a text file with a list of paths to fasta files)
+ref = snakemake.input["ref"]
+if ref.endswith((".fa", ".fna", ".fasta")):
```
Comment (fgvieira) [inline bio/mashmap/wrapper.py:18]: Does `MashMap` support gzipped reference genomes?

**Comment by fgvieira** [comment]
It seems that `MashMap` has its own output format, "space-delimited with each line consisting of query name, length, 0-based start, end, strand, target name, length, start, end and mapping nucleotide identity".

Just for the sake com consistency with other read mappers, is there a way to convert this output to a SAM/BAM/CRAM?

File: bio/mashmap/wrapper.py (hunk 3 above)
Comment (tdayris) [reply bio/mashmap/wrapper.py:18]: Unfortunately not. Do you think this wrapper should support on-the-fly decompression of reference fasta files ?

**Comment by tdayris** [comment]
> Just for the sake com consistency with other read mappers, is there a way to convert this output to a SAM/BAM/CRAM?

Mashmap format to SAM/BAM/CRAM is not supported by MashMap. There is an open issue about that on their git. I'm wrapping this tool to follow Salmon guideline to create a gentrome file with decoy sequences.

> Also, does MasMap uses any temp folder?

There is no official variable or command line argument to overload any default temporary directory.

File: bio/mashmap/meta.yaml [hunk 4]
```diff
@@ -11,7 +11,5 @@ output:
 params:
   - extra: Optional parameters for MashMap
 notes: |
-  `input.ref` may be either a path to a file containing sequences (fasta) or a text file containing a list
-  paths to fasta-formatted sequence files.
-  
-  `input.query` may be either a fastq file or a fasta file.
\ No newline at end of file
+  * `input.ref` may be either a path to a fasta file or a text file containing a list of paths to several fasta files.
+  * `input.query` may be either a fastq file or a fasta file.
```
Comment (fgvieira) [inline bio/mashmap/meta.yaml:15]: According to your latest changes, it seems this one can also be a text file with a list of FASTA/Q files, no?

File: bio/mashmap/wrapper.py (hunk 3 above)
Comment (tdayris) [reply bio/mashmap/wrapper.py]: Okay, I'm still on this. Using `<(gzip ...)` syntax leads to a 127 error. Command not found. But gzip is here, mashmap and gsl are here ... The error only occurs with gzipped fasta files as reference.

Comment (tdayris) [reply bio/mashmap/wrapper.py]: Sooo ... It works here ... It does not locally.

**Review by fgvieira — APPROVED** [review]

**Pull request merged**